File: solutions/linked-list/019-remove-nth-node-from-end-of-list.js
I'll help you create a complete JavaScript implementation for removing the Nth node from a linked list. I'll write the solution with clear comments and proper error handling.

```javascript
/**
 * Definition for singly-linked list node
 */
class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

/**
 * Removes the nth node from the end of a linked list
 * @param {ListNode} head - The head of the linked list
 * @param {number} n - The position from the end to remove (1-based)
 * @return {ListNode} - The head of the modified linked list
 */
function removeNthFromEnd(head, n) {
    // Handle edge cases
    if (!head) return null;
    if (n <= 0) return head;

    // Create a dummy node to handle edge cases (like removing first node)
    const dummy = new ListNode(0);
    dummy.next = head;
    
    // Use two pointers: fast and slow
    let fast = dummy;
    let slow = dummy;
    
    // Move fast pointer n steps ahead
    for (let i = 0; i <= n; i++) {
        if (!fast) return head; // Handle case where n is larger than list length
        fast = fast.next;
    }
    
    // Move both pointers until fast reaches the end
    while (fast) {
        fast = fast.next;
        slow = slow.next;
    }
    
    // Remove the nth node by updating the next pointer
    slow.next = slow.next ? slow.next.next : null;
    
    return dummy.next;
}

/**
 * Helper function to create a linked list from an array
 * @param {Array} arr - Array of values to create linked list from
 * @return {ListNode} - Head of the created linked list
 */
function createLinkedList(arr) {
    if (!arr.length) return null;
    
    const head = new ListNode(arr[0]);
    let current = head;
    
    for (let i = 1; i < arr.length; i++) {
        current.next = new ListNode(arr[i]);
        current = current.next;
    }
    
    return head;
}

/**
 * Helper function to convert linked list to array for testing
 * @param {ListNode} head - Head of the linked list
 * @return {Array} - Array representation of the linked list
 */
function linkedListToArray(head) {
    const result = [];
    let current = head;
    
    while (current) {
        result.push(current.val);
        current = current.next;
    }
    
    return result;
}

// Test cases
function runTests() {
    // Test Case 1: Regular case
    let test1 = createLinkedList([1, 2, 3, 4, 5]);
    console.log('Test 1:', linkedListToArray(removeNthFromEnd(test1, 2)));  // Should print [1,2,3,5]
    
    // Test Case 2: Remove first element
    let test2 = createLinkedList([1, 2]);
    console.log('Test 2:', linkedListToArray(removeNthFromEnd(test2, 2)));  // Should print [2]
    
    // Test Case 3: Remove last element
    let test3 = createLinkedList([1, 2, 3]);
    console.log('Test 3:', linkedListToArray(removeNthFromEnd(test3, 1)));  // Should print [1,2]
    
    // Test Case 4: Single element
    let test4 = createLinkedList([1]);
    console.log('Test 4:', linkedListToArray(removeNthFromEnd(test4, 1)));  // Should print []
    
    // Test Case 5: Empty list
    console.log('Test 5:', linkedListToArray(removeNthFromEnd(null, 1)));  // Should print []
}

// Run the tests
runTests();

// Export the functions for external use
module.exports = {
    removeNthFromEnd,
    ListNode,
    createLinkedList,
    linkedListToArray
};
```

This implementation includes:

1. A `ListNode` class for creating linked list nodes
2. The main `removeNthFromEnd` function that removes the nth node from the end
3. Helper functions for creating linked lists and converting them to arrays
4. A comprehensive test suite
5. Proper error handling for edge cases
6. Clear comments explaining the implementation
7. Module exports for external use

The solution uses the two-pointer technique to find the nth node from the end in a single pass. The algorithm:

1. Uses a dummy node to handle edge cases
2. Maintains two pointers (fast and slow) separated by n nodes
3. When the fast pointer reaches the end, the slow pointer is at the correct position to remove the nth node
4. Updates the pointers to skip the nth node

The time complexity is O(n) where n is the length of the linked list, and the space complexity is O(1) as we only use a constant amount of extra space.

The test cases cover various scenarios including:
- Regular case with multiple nodes
- Removing first element
- Removing last element
- Single element list
- Empty list

File: solutions/linked-list/019-remove-nth-node-from-end-of-list.py
"""
# 019. Remove Nth Node From End Of List
**Medium**

Given a problem that demonstrates key concepts in Linked List.

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of linked list concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply linked list methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages linked list principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### TIME COMPLEXITY: O(n)
### SPACE COMPLEXITY: O(1)

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>

<details>
<summary><b>💡 APPROACH</b></summary>

The approach uses linked list techniques to solve this problem efficiently.

### Algorithm Steps:
1. Initialize necessary variables
2. Process input using linked list method
3. Return the computed result

</details>
"""

class ListNode:
    """Definition for singly-linked list."""
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        """
        Remove the nth node from the end of the list using two-pointer technique.

        Args:
            head: Head of the linked list
            n: Position from the end to remove (1-indexed)

        Returns:
            Head of the modified linked list

        Time Complexity: O(n) - single pass through the list
        Space Complexity: O(1) - only using two pointers
        """
        # Create a dummy node to handle edge case where head is removed
        dummy = ListNode(0)
        dummy.next = head

        # Initialize two pointers
        fast = slow = dummy

        # Move fast pointer n+1 steps ahead
        for _ in range(n + 1):
            fast = fast.next

        # Move both pointers until fast reaches the end
        while fast:
            fast = fast.next
            slow = slow.next

        # Remove the nth node from end
        slow.next = slow.next.next

        return dummy.next

    def solve(self, head: ListNode, n: int) -> ListNode:
        """Wrapper method for consistency with template."""
        return self.removeNthFromEnd(head, n)

def list_to_array(head: ListNode) -> list:
    """Convert linked list to array for testing."""
    result = []
    while head:
        result.append(head.val)
        head = head.next
    return result

def array_to_list(arr: list) -> ListNode:
    """Convert array to linked list for testing."""
    if not arr:
        return None
    head = ListNode(arr[0])
    current = head
    for val in arr[1:]:
        current.next = ListNode(val)
        current = current.next
    return head

def test_solution():
    """
    Test cases for 019. Remove Nth Node From End Of List.
    """
    solution = Solution()

    # Test case 1: Remove from middle
    head = array_to_list([1, 2, 3, 4, 5])
    result = solution.solve(head, 2)
    expected = [1, 2, 3, 5]
    assert list_to_array(result) == expected, f"Expected {expected}, got {list_to_array(result)}"

    # Test case 2: Remove head (single node)
    head = array_to_list([1])
    result = solution.solve(head, 1)
    expected = []
    assert list_to_array(result) == expected, f"Expected {expected}, got {list_to_array(result)}"

    # Test case 3: Remove head (multiple nodes)
    head = array_to_list([1, 2])
    result = solution.solve(head, 2)
    expected = [2]
    assert list_to_array(result) == expected, f"Expected {expected}, got {list_to_array(result)}"

    # Test case 4: Remove last node
    head = array_to_list([1, 2, 3])
    result = solution.solve(head, 1)
    expected = [1, 2]
    assert list_to_array(result) == expected, f"Expected {expected}, got {list_to_array(result)}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    head = array_to_list([1, 2, 3, 4, 5])
    result = solution.solve(head, 2)
    print(f"Solution for 019. Remove Nth Node From End Of List: {list_to_array(result)}")
